tests/test_context_local.py

```diff
@@ -0,0 +1,194 @@
+import asyncio
+import threading
+import time
+import typing
+import unittest
+from collections import Counter
+
+from opnieuw.retries import retry_async, retry
+from opnieuw.test_util import retry_immediately
+from opnieuw.util import no_retries
+from tests.utils import AsyncTestCase
+
+
+# The barriers are used to ensure all asyncio tasks/threads have their retry state modified
+# even while the other asyncio tasks/threads are still running.
+# Otherwise, one asyncio task/thread could finish and revert the retry state before the other starts.
+# We want to test the retry state context during concurrency, so the flow should be:
+#
+#        T1         |       T2
+# ------------------|------------------
+# modify state      |
+#                   | modify state
+# wait for others   | wait for others
+# perform test      | perform test
+#                   |
+# wait for others   | wait for others
+# revert state      | revert state
+#
+# This ensures that if the retry state context is leaked across asyncio task or thread,
+# the last one to modify the state will make the others' test fail.
+
+
+class AsyncBarrier:
+    def __init__(self, n: int) -> None:
+        self.n = n
+        self._current = 0
+        self.event = asyncio.Event()
+
+    async def wait(self) -> None:
+        self._current += 1
+        if self._current == self.n:
+            self.event.set()
+        await self.event.wait()
+
+
+class TestAsyncContext(AsyncTestCase):
+    counter: typing.Counter[str] = Counter()
+
+    @retry_async(
+        retry_on_exceptions=TypeError,
+        max_calls_total=3,
+        retry_window_after_first_call_in_seconds=3,
+    )
+    async def async_foo(self, counter_key: str) -> None:
+        self.counter[counter_key] += 1
+        # Yield control back to the event loop to give other coro a chance to run
+        await asyncio.sleep(0)
+        raise TypeError
+
+    def test_async_retry_state_context(self) -> None:
+        """
+        Test that the retry state is only modified in the context of an asyncio.Task
+        and not globally.
+        """
+
+        async def _retry_immediately_coro(
+            start_barrier: AsyncBarrier, end_barrier: AsyncBarrier
+        ) -> None:
+            counter_key = "retry_immediately"
+
+            with retry_immediately():
+                await start_barrier.wait()
+                start = time.monotonic()
+                try:
+                    await self.async_foo(counter_key)
+                except TypeError:
+                    end = time.monotonic()
+                    runtime_seconds = end - start
```

Comment:
In theory the things could change that would cause `async_foo` to not finish on time. In those cases the tests would spend a bit of additional time before failing. You could force a success limit using `await asyncio.wait_for(self.async_foo(counter_key), timeout=0.5)`, that should make the test fail a little quicker.